### ISSUE TITLE
get error 'A branch with that name already exists.' when trying to access 'other branches'

### ISSUE THREAD

**Issue by ggrrll** [desktop/desktop#4527]
Hi, 

first of all, thanks for this nice tool.

I am getting this error message `A branch with that name already exists.` , when I click on `Other branches` , which are branches that do not appear anyway with `git ls-remote` – not sure if this is related to #2964 

Thanks for your help.

**Comment by shiftkey** [comment]
@ggrrll as mentioned in #3624, I think there's a corner case here that's tricky to fix:

https://github.com/desktop/desktop/blob/be2712877593641de83b7104da6569e113a2f033/app/test/unit/git/checkout-test.ts#L112

This test tries to checkout a remote branch where it has the same name as an existing local branch, and it currently fails. Can you confirm that's the same behaviour you're seeing:

 - you have a existing local branch named [something]
 - you're trying to checkout a remote branch named [something]

**Comment by ggrrll** [comment]
Thanks @shiftkey.

Not sure indeed of what's happening... :( 
I guess I messed up a bit in the past, as I try to connect two other remote repos (actually from [Overleaf](https://www.overleaf.com))  to the same local directory, which I later removed.

They way I understand it is that those two 'other branches' are old remote repositories which do not exist anymore locally (indeed, are not in the output of `git ls-remote` )... so, I don't see why github desktop is still showing them...

**Comment by shiftkey** [comment]
Could you attach the full log files? The **Help** | **Show Logs in Explorer** menu items should open the logs folder, and inside those are the log files generated by day. This will provide some more information to help troubleshoot.

**Comment by ggrrll** [comment]
sure – this is the part of the error

```
2018-04-25T12:43:38.725Z - error: [ui] `git -c credential.helper= checkout --progress overleaf/master -b master --` exited with an unexpected code: 128.
fatal: A branch named 'master' already exists.

(The error was parsed as 25: A branch with that name already exists.)
```

**Comment by iAmWillShepherd** [comment]
@ggrrll could you attach  the **full* logs?

**Comment by ggrrll** [comment]
ok sorry, here you go
[2018-04-25.desktop.production.log](https://github.com/desktop/desktop/files/1947837/2018-04-25.desktop.production.log)

**Comment by Kherae** [comment]
I'm having a similar issue to this (possibly the same, terminology is still new to me), though with forks. I've a separate fork from a project, rather than a branch. I get the same error when trying to switch to upstream/master.

**Comment by shiftkey** [comment]
> A potential solution for this would be to, in cases where an existing local branch exists and will clash, use the convention `{remote}-{branch}` when the user wants to checkout the same branch from another remote.

Taking this a step further, rather than introducing a magic convention here would it be helpful to explain the problem and let the user choose what to name the branch (with a sane default)?

![](https://user-images.githubusercontent.com/359239/39566006-72eb7db6-4efd-11e8-8777-8ab42ac3c6e7.gif)

The words in here are just something I threw together in haste, but I wanted to demonstrate the flow that might help address the bug and better support users working with multiple remotes.

**Comment by theFisher86** [comment]
Having the same issue... surprised there's not a fix here

**Comment by TizianoNoto** [comment]
> 
> 
> I'm having a similar issue to this (possibly the same, terminology is still new to me), though with forks. I've a separate fork from a project, rather than a branch. I get the same error when trying to switch to upstream/master.

I confirm. I have a fork project and I can't checkout "upstream/master" because of this issue. I have to open Git GUI to see what's changed.

**Comment by shiftkey** [comment]
Adding `user-research` to this label as we got a few 👍 to my suggestion above from a few months ago but unclear on next steps.

**Comment by danilaml** [comment]
I've just moved from GitHub for Windows app and this issue is really frustrating to me. In GH4W there was no such problem. You could see `upstream/master` branch and switch/checkout it without any problem. It'd be "checked out" as exactly this, `upstream/master`,  no issue here, everything works as expected.

I liked and used this feature a lot, because that way you could easily develop in your own fork, branching from upstream/master, without all the pain of keeping your master in sync with the fork.
When I needed, I could just check out `remotename/branchname`, branch from this and continue working.

GitHub for Desktop completely breaks this workflow for me for no apparent reason.

**Comment by shiftkey** [comment]
@danilaml would the solution I talked about here https://github.com/desktop/desktop/issues/4527#issuecomment-386221312 help address this limitation?

**Comment by shiftkey** [comment]
I'll also point out that we're tracking an upstream Git limitation about how to correctly handle remote branch names containing `/` - #3618 - which is why I recommended using `-` here rather than `/`. But given we use `pr/{number}` refs to track pull request branches (and can push to those) I think either approach is fine.

**Comment by danilaml** [comment]
@shiftkey If that branch's contents would be automatically synced up with the remote, and GitHub for desktop wouldn't try to push it into my fork (unless specifically asked to), I could live with that.

With GH4W, the workflow felt like you never actually "checked out" remote branch, i.e. you've just viewed it's remote state without actually modifying anything in your local repo, if that makes sense.
Basically, I want a way to seamlessly view the current state of the remote and sometimes branch from it.

Btw, was there a discussion regarding ` git status --branch` (as indicated by the last comment in linked issue)?

**Comment by shiftkey** [comment]
@danilaml I dusted off the old GHfW code and tested this out with my fork of Desktop, and we do create a new branch that matches the same ref (and tracking the right remote):

```shellsession
$ git status
On branch desktop/master
Your branch is up to date with 'remotes/desktop/master'.

nothing to commit, working tree clean
```

We could follow that pattern here too, but this bit

> wouldn't try to push it into my fork (unless specifically asked to)

won't happen because of how Git sets up the default branches  - you'd have to merge changes into `origin/master` to then push them up to the remote for that branch, and we don't currently support pushing arbitrary refs to arbitrary remotes.

> Btw, was there a discussion regarding `git status --branch (as indicated by the last comment in linked issue)?

I'm not sure I understand - the discussion hasn't moved beyond that last comment because it requires a contribution to upstream Git (I'd love to get to it, but Priorities™).

**Comment by danilaml** [comment]
>won't happen because of how Git sets up the default branches

Well, that's fine. I just don't want for this pseudo-remote-branches to show up in my GH fork without me knowing. I.e. I'm fine with them being local, just not created on my origin automatically.

>I'm not sure I understand 

Well, the comment also talked about searching for existing discussions/asking mailing list so I'd like to follow/read that if it's been done.

**Comment by danilaml** [comment]
Another thing that'd greatly simplify things (and cover 95% of my use cases with remote branches) if I was able to select some other branch from this dialog (new branch):

![image](https://user-images.githubusercontent.com/2046350/48306503-d04f5200-e54a-11e8-80eb-3bab3148a4a8.png)

I can't see the reason for why there are only 2 choices. There should be a way to select arbitrary branch that you can see in branches drop-down. And that way, I'd be able to branch from remote/branch without actually checking it out into some separate local branch (I did it via command line so I don't see why It can be added to the interface).

**Comment by kuychaco** [comment]
I explored this a bit and decided to put it down for now in favor of other work. I'll do a brain dump here in order to capture what I thought about and discovered.

### Proposed solution
Say we've forked a repo, open it in Desktop (using the "Clone or download" button), and click "Fetch origin" to ensure that the remote branch refs have been fetched. 

In the branch list we see `master` which tracks with `origin/master`. And under "Other Branches" we see `upstream/master`. 

Right now, this `upstream/master` is not actually referring to a local branch. It's referring to a **remote branch** that Git uses to track where the remote branch tip is (presumably output from `git branch --remote`). 

Currently when the user clicks this Desktop tries to create another branch called `master`, which fails because we of course already have a branch called `master`.

Ideally, when the user clicks this, Desktop will detect that there's a clash in branch name and create a new local branch called `upstream/master` and set up the proper remote tracking info for this branch. 

Under the hood, Desktop could run `git checkout -b upstream/master remotes/upstream/master` and this would produce the following config settings:

```
[branch "upstream/master"]
	remote = upstream
	merge = refs/heads/master
```

Now when the user checks the branch list they should see `upstream/master` in their "Recent branches" list. 

### Considerations

**Note**: Currently if I run `git checkout -b upstream/master remotes/upstream/master` on the command line, for some reason Desktop displays `heads/upstream/master` in the "Other Branches" list, and when clicked on the app puts you in detached head state. It’s not immediately clear to me why this happens. Interestingly if I name the branch upstream-master then it shows up just fine and works as expected.

Another consideration to keep in mind is -- if a user clicks a branch corresponding to an upstream remote branch that doesn't have a local branch name that conflicts, Desktop currently creates a local branch for it that does not have any `upstream` prefix. This inconsistency could be confusing... Here's an example to illustrate: say we have branches on our upstream remote called `master` and `giraffe`. If a user clicks the `upstream/master` remote branch in the list then, because there's already a local `master` branch corresponding to `remotes/origin/master`, they'd end up with a `upstream/master` local branch that tracks the `remotes/upstream/master` branch. If they click `upstream/giraffe` and there is not already a local branch with that name, they'll get a `giraffe` local branch that tracks the `remotes/upstream/giraffe` branch. 

Also worth noting, the Desktop team discussed prompting the user for the branch name they would like to use when there's a naming conflict (per @shiftkey's suggestion [here](https://github.com/desktop/desktop/issues/4527#issuecomment-386221312)). We are leaning towards a more behind-the-scenes approach in order to spare users potential confusion if they're unaware of what the clash means or why it's happening. And we imagine that a vast majority of cases where this is an issue, a user is using a forking workflow and trying to see what's on `upstream/master`, so we can simply create a new branch called `upstream/master` or them to be able to do so (but of course generalize this to `<remote-name>/<branch-name>`). This would spare users some unnecessary mental cycles, and perhaps spare newer folks some confusion or intimidation.

**Comment by muralib91dotnet** [comment]
> Currently I use the following workaround: switch to console/shell and checkout `remote/branch` manually (that works fine), when branch from that. Hopefully it'll be automated soon.

Worked for me

**Comment by csrvermaak** [comment]
@kuychaco - thanks for perfectly describing the problem. I tried your solution, but unfortunately didn't get the required result.:


_Under the hood, Desktop could run 'git checkout -b upstream/master remotes/upstream/master` and this would produce the following config settings:_

```
[branch "upstream/master"]
	remote = upstream
	merge = refs/heads/master

```

_Now when the user checks the branch list they should see upstream/master in their "Recent branches" list._

---

What I got after running 'git checkout -b upstream/master remotes/upstream/master' was:

```
Switched to a new branch 'upstream/master'
Branch 'upstream/master' set up to track remote branch 'master' from 'upstream'.

```

This sounded promising. As you mentioned, the 'config' file was updated to include:

```
[branch "upstream/master"]
	remote = upstream
	merge = refs/heads/master

```


However, when I opened GitHub Desktop, I got the following:

![image](https://user-images.githubusercontent.com/6738800/71681510-961bc200-2d95-11ea-8af5-549e79663a7b.png)


And when I try to switch to 'heads/upstream/master', I get a detached head:

![image](https://user-images.githubusercontent.com/6738800/71681580-b77cae00-2d95-11ea-9539-aa9a25746e32.png)


Hope you can provide some guidance? I've manually deleted the new branch in the interim. Thanks.

**Comment by bls1999** [comment]
> This sounded promising. As you mentioned, the 'config' file was updated to include:
> 
> ```
> [branch "upstream/master"]
> 	remote = upstream
> 	merge = refs/heads/master
> ```
> 
> However, when I opened GitHub Desktop, I got the following:
> 
> ![image](https://user-images.githubusercontent.com/6738800/71681510-961bc200-2d95-11ea-8af5-549e79663a7b.png)
> 
> And when I try to switch to 'heads/upstream/master', I get a detached head:
> 
> ![image](https://user-images.githubusercontent.com/6738800/71681580-b77cae00-2d95-11ea-9539-aa9a25746e32.png)
> 
> Hope you can provide some guidance? I've manually deleted the new branch in the interim. Thanks.

This is happening to me too. Any guidance would be much appreciated.

**Comment by steveward** [comment]
From #10726:

> When working with forks and PRs in general, I often have a branch that has the same name in the fork (origin) versus the original (upstream). I prefer to base all of my PRs off of the upstream variant of the branch (e.g. master), but I'm unable to check it out since I have a branch with the same name on my fork.
> 
> To avoid dirty PRs and such, I find myself going to the command line to forcibly change the upstream branch so that I don't have this problem anymore, but it would be nice to be able to both see and change the upstream branch within the UI.
> 
> Proposed solution
> Add new option in the "Branch" dropdown menu called "Change Tracking Branch". If that's too complex, just the ability to change the remote (via "Change Remote") or similar would still satisfy my needs.
>

**Comment by Starli0n** [comment]
I was searching for good git client for my teammates.

So I gave a try to Desktop Github, to discover that I could not switch to my upstream whereas it is thanks to github that I learnt git by forking and having an upstream.

Seriously, the bug was opened since almost 3 years and it is not resolved: _SHAME !!_
whereas it is a basic github workflow, this is a joke 😂 

I do not care about command lines workaround as I am searching a UI client.
Since I know git, I am waiting for a decent git UI but it seems that it is _not today_

Less wait again few years...

**Comment by TizianoNoto** [comment]
> I do not care about command lines workaround as I am searching a UI client.
> Since I know git, I am waiting for a decent git UI but it seems that it is _not today_

I've always copied the repositories with GitHub Desktop and then, because of this ~feature~ bug, used Git's built-in GUI tools for updates from upstream and to check mistakes before pushing.

@Starli0n I guess you already did but, if not, take a look at other [suggested third-party GUI clients](https://git-scm.com/downloads/guis) and see if any of them fits your needs.

**Comment by Starli0n** [comment]
> @Starli0n I guess you already did but, if not, take a look at other [suggested third-party GUI clients](https://git-scm.com/downloads/guis) and see if any of them fits your needs.

Thank you 

Indeed, this is where I started as it seems to be ordered by popularity.

As I am using VSCode, I will give a try to use Git lens plugin and use an integrated version instead a third party tool.

Anyway, both are microsoft now...

**Comment by vurvdev** [comment]
Pls..

**Comment by Fasteroid** [comment]
Pls..

**Comment by shiznit304** [comment]
its been a couple of years and still no fix

**Comment by ArjixWasTaken** [comment]
Please fix this already.

**Comment by Kherae** [comment]
I am...sorely disappointed. Over three years and still unfixed says it is never going to get fixed, and this bug is a very big annoyance when working on some team projects.

**Comment by jintsi** [comment]
Can this get fixed already?

**Comment by t3chn0m4g3** [comment]
This still seems to be unresolved 😢

**Comment by zeusongit** [comment]
Waiting...

**Comment by vu1canic** [comment]
Still waiting, I'm hoping for a fix

**Comment by shiznit304** [comment]
You can merge a different way by comparing changes feature

**Comment by Fasteroid** [comment]
dear god will this ever get fixed, it's been 4 years

**Comment by ArjixWasTaken** [comment]
![image](https://user-images.githubusercontent.com/53124886/166163639-5a876be4-5367-47d5-949d-92cb67ad0bca.png)

> Doesn't prevent core functions

I think you really need to reconsider what "core functions" are...
Checking out a remote branch is a major function, and it should be high priority.

**Comment by quadrismegistus** [comment]
Just ran into this. +1 please fix thank you

**Comment by prateekgangopadhyay** [comment]
Same issue. Please share the updates! +2 (one for my colleague)

**Comment by jhallbachner** [comment]
I have avoided using a GUI git client for years as I feel very familiar with the CLI interface on *nix systems, but I'm currently updating a repo (a fork of another user's primary origin) on a Windows system and thought I would give this official client a try in order to manage this repo here. Running into this bug within thirty seconds of installing the client and then discovering it hasn't been fixed for over four years leaves an appalling first impression. Software with bugs of this severity, regularly encountered by users following bog-standard workflows, should be considered abandoned and it's inappropriate to promote them as actively supported software on the site.

**Comment by danielniccoli** [comment]
@ArjixWasTaken is right. I cannot check out upsteam/dev, so I'm unable to merge one of the most important branches of the original repo that I am contributing to with the client. That is pretty core to me.

**Comment by shiznit304** [comment]
There’s a work around. You can compare changes of your current branch to
upstream and merge through that.

**Comment by danielniccoli** [comment]
@shiznit304 Thank you, this helps immensely!

**Comment by Syntex9000** [comment]
Still a problem when using a fork. Can't find a work around to that. This GUI is basically useless to me

**Comment by shiznit304** [comment]
Use the work around

**Comment by aejontargaryen** [comment]
Used the workaround mentioned by @shiznit304 
Its in top toolbar of GUI under "**Branch**"  
then "_compare to branch_"

**Comment by samuelko123** [comment]
Hi @steveward (or any maintainer),

I'm hoping to give it a crack.

There're currently two proposed solutions
1. [Display naming conflict](https://github.com/desktop/desktop/issues/4527#issuecomment-386221312)
2. [Git checkout magic](https://github.com/desktop/desktop/issues/4527#issuecomment-523660277)

I just tested solution 2 and it failed, same outcome as [this comment](https://github.com/desktop/desktop/issues/4527#issuecomment-570278344).

Also, running git commands would result in the same "detached HEAD" issue:
![image](https://user-images.githubusercontent.com/91960206/234251555-496c3193-bf6e-4128-b295-b4b664f20c8f.png)

## Proposed Solution
It'd be based on [solution 1](https://github.com/desktop/desktop/issues/4527#issuecomment-386221312).
The concern is "we will confuse new users" by display "name clash".

The solution is better wording.
![image](https://user-images.githubusercontent.com/91960206/234256407-9a549637-2458-443c-a769-0cefe28ef244.png)

Given the "popularity" of this bug, I'd say having something is better than hoping for an ideal solution which never exists.
Hope it makes sense.

**Comment by noskill24** [comment]
Recently downloaded this GUI, and now I'm here because of this issue. 
WOW! 6 years still a problem when using multiple remotes?
Is this project abandoned?

**Comment by ArjixWasTaken** [comment]
> Recently downloaded this GUI, and now I'm here because of this issue. 
> WOW! 6 years still a problem when using multiple remotes?
> Is this project abandoned?

Use Fork or GitKraken instead :rofl:

**Comment by Zuzu-Typ** [comment]
This proposed solution should be fairly easy to implement.
It's not a good idea to keep people waiting for this long.
You really should address this issue soon.

> > A potential solution for this would be to, in cases where an existing local branch exists and will clash, use the convention `{remote}-{branch}` when the user wants to checkout the same branch from another remote.
> 
> Taking this a step further, rather than introducing a magic convention here would it be helpful to explain the problem and let the user choose what to name the branch (with a sane default)?
> 
> ![ ](https://user-images.githubusercontent.com/359239/39566006-72eb7db6-4efd-11e8-8777-8ab42ac3c6e7.gif)
> 
> The words in here are just something I threw together in haste, but I wanted to demonstrate the flow that might help address the bug and better support users working with multiple remotes.

**Comment by T9Air** [comment]
Will this be fixed?

**Comment by Yolwoocle** [comment]
Hey, it's been 8 years, is there any plans to fix this?
